### ISSUE TITLE
hpack all packages in a project with one invocation

### ISSUE THREAD

**Issue by quasicomputational** [sol/hpack#347]
Currently, hpack is totally project-agnostic. If you're working in a project with multiple packages and you want to regenerate many or even all of the `.cabal` files, you have to invoke hpack for each package. For cabal-install projects, that can be quite the hardship; it's OK for stack because of the built-in hpack integration (i.e., `stack build` will regenerate the package descriptions for you).

I think it'd be nice to have a hpack mode that'd re-run hpack for all the packages in the project. This would need a `--project-file` flag to point hpack at the `.project` file, if it's in a non-standard place. There's also the problem that `cabal.project` files don't have a parser we can use. (I do have some code for this but it'd need some polishing before release.)

**Comment by sol** [comment]
How about `find -name package.yaml -exec hpack {} \;`?

**Comment by sol** [comment]
Not eager to add `cabal` specific features, but we could consider to do the `find` equivalent on e.g. --all.

**Comment by quasicomputational** [comment]
One problem with that is that it'd not work if you're in a subdirectory. cabal-install and stack both go up the directory tree until they find a project description and then they'll do the right thing, but presumably `hpack --all` wouldn't.

**Comment by sol** [comment]
@quasicomputational if a proper solution to this requires to be aware of `cabal.project` the do you think having this in a separate executable/project would be an option?

**Comment by JonathanILevi** [comment]
I think that the best solution would be to be cabal.project aware.  Using hpack on only and all the packages in cabal.project (or more ideally some way to specify which ones).  Maybe a `project.yaml`?  Either way I think a separate executable would be fine (hproj? :smile:)

**Comment by sol** [comment]
@quasicomputational did you do anything here?  I would want to see a proof-of-concept of this in a separate project / executable.

**Comment by quasicomputational** [comment]
No, I never wound up writing anything to do this that would've been anywhere near suitable for general consumption - all entangled with other considerations.

**Comment by philderbeast** [comment]
I use a fairly simple shake build script for [generating cabal files in a project](https://github.com/BlockScope/ghc-corroborate/blob/main/build/Main.hs). It uses `package.dhall` files instead of `package.yaml`. I build this project and invoke with a bash script but maybe it could be done with `runhaskell`?
